Decouple knot_cloud from amqp calls

agms-cesar opened the pull request:
- Removes amqp direct calls from knot_cloud;
- Creates an unique publish message method in mq;
- Creates an enum to identify mq message type;
- Encapsulates mq messages in a struct mq_message_data_t;
- Removes unused queue_reply;
- Moves queue_fog to mq.c with a new name worker_queue;
- Moves mq message header to mq.c;
- Modifies mq.c to deal properly with worker_queue;
- Refactors knot_cloud to use mq_message_data_t struct.

CaioCordeiro commented on src/knot_cloud.c:
```diff
@@ -297,23 +263,22 @@ static int create_cloud_queue(const char *id)
 
 	snprintf(queue_fog_name, sizeof(queue_fog_name), "%s-%s",
 		 MQ_QUEUE_FOG_OUT, id);
-
```
Is removing this line necessary?

agms-cesar replied:
Actually no, i think i removed by mistake

agms-cesar replied:
done

CaioCordeiro commented on src/knot_cloud.c:
```diff
@@ -407,9 +361,7 @@ int knot_cloud_register_device(const char *id, const char *name)
-
-	headers[0].value.value.bytes = amqp_cstring_bytes(user_auth_token);
-
+	
```
It would be interesting to just leave the line as it was

agms-cesar replied:
> I've made some change requests in respect to comments and blank lines, but the code looks good!
> If it is possible, I would just consider splitting this commit in more than one commit.
> E.g: One to effectively decouple the knot_cloud from amqp calls, another one to Removes unused queue_reply;
> Great Pull Request!

I agree with you, i will do this

CaioCordeiro commented on src/mq.c:
```diff
@@ -603,11 +559,10 @@ int mq_delete_queue(amqp_bytes_t queue)
  * Returns: 0 if successful and -1 otherwise.
  */
-int mq_consumer_queue(amqp_bytes_t queue)
+int mq_consumer_queue(const char *name)
```
Update the parameters in the comment above.

CaioCordeiro commented on src/mq.c:
```diff
@@ -655,26 +610,29 @@ int mq_set_read_cb(mq_read_cb_t read_cb, void *user_data)
 	mq_ctx.conn_retry_timeout = l_timeout_create_ms(1, // start in oneshot
 							attempt_connection,
 							url, NULL);
-
```
Is removing this line necessary?

CaioCordeiro reviewed (changes requested):
I've made some change requests in respect to comments and blank lines, but the code looks good!
If it is possible, I would just consider splitting this commit in more than one commit.
E.g: One to effectively decouple the knot_cloud from amqp calls, another one to Removes unused queue_reply;
Great Pull Request!

vitbaq reviewed (changes requested):
LGTM, great patch.
I think that you can break this single commit into several. This makes it easier for the review and the use of the git functionalities.

agms-cesar committed:
Removes mq defines and unused queue

- Moves mq defines to mq.h;
- Removes queue_reply which is not used in knot_cloud;

agms-cesar committed:
Refactors knot_cloud to use new mq message struct

- Removes mq message header definition from knot_cloud;
- Adds mq message header definition to mq.c;
- Creates an unique publish message method in mq;
- Refactors knot_cloud to use new unique publish method;
- Creates a define to number of headers in mq.c;
- Adds user_token parameter to mq_start in mq.c;
- Modifies mq_start to create mq message header.

vitbaq commented on src/mq.c:
```diff
@@ -523,53 +522,48 @@ amqp_bytes_t mq_declare_new_queue(const char *name)
- * Returns: 0 if successful and -1 otherwise.
  */
-int mq_delete_queue(amqp_bytes_t queue)
+void mq_delete_queue()
```
Put void as a parameter

vitbaq commented on src/mq.c:
```diff
@@ -523,53 +522,48 @@ amqp_bytes_t mq_declare_new_queue(const char *name)
  * Returns: 0 if successful and -1 otherwise.
  */
-int mq_consumer_queue(amqp_bytes_t queue)
+int mq_consumer_queue()
```
Put void as a parameter

vitbaq commented on src/mq.h:
```diff
@@ -84,14 +84,12 @@ typedef void (*mq_connected_cb_t) (void *user_data);
+int mq_prepare_direct_queue(const char *name, const char *exchange,
+			 const char *routing_key);
+int mq_declare_new_queue(const char *name);
+void mq_delete_queue();
+int mq_consumer_queue();
```
Put void as a parameter

CaioCordeiro commented on src/mq.c:
```diff
@@ -561,53 +522,48 @@ amqp_bytes_t mq_declare_new_queue(const char *name)
+	worker_queue = amqp_bytes_malloc_dup(r->queue);
+	if (worker_queue.bytes == NULL)
 		l_error("Out of memory while copying queue buffer");
-
```
Is removing this line necessary?

agms-cesar replied:
No, it isn't. I will fix this

CaioCordeiro commented on src/mq.c:
```diff
@@ -561,53 +522,48 @@ amqp_bytes_t mq_declare_new_queue(const char *name)
+		amqp_bytes_free(worker_queue);
+		worker_queue = amqp_empty_bytes;
 	}
-
```
Is removing this line necessary?

agms-cesar replied:
Yes, otherwise there will be two blank lines

agms-cesar replied:
Yes, because the return statement was removed, so this is the method's last line (a curly bracket)

CaioCordeiro reviewed (changes requested):
Some change requests that passed me in the past review.

vitbaq commented on src/mq.c:
```diff
@@ -54,6 +56,9 @@ struct mq_context {
 static struct mq_context mq_ctx;
+static const int8_t num_of_headers = MQ_NUM_OF_HEADERS;
+amqp_table_entry_t headers[MQ_NUM_OF_HEADERS];
+amqp_bytes_t worker_queue;
```
Idk if it is right to call this worker_queue. A worker queue is intended to distribute tasks among workers (https://www.rabbitmq.com/tutorials/tutorial-two-python.html) and it is not what happens here. Maybe we can just call as queue or current_queue.

CaioCordeiro reviewed (approved):
LGTM

CaioCordeiro reviewed (approved):
LGTM.

alla3 commented on src/knot_cloud.c:
```diff
@@ -421,11 +373,13 @@ int knot_cloud_register_device(const char *id, const char *name)
-					   json_str);
+	mq_message_data_t mq_message = {
+				 MQ_MESSAGE_TYPE_DIRECT, MQ_EXCHANGE_DEVICE,
+				 MQ_CMD_DEVICE_REGISTER, MQ_MSG_EXPIRATION_TIME_MS, json_str,
```
We try to keep the code maximum column number at 80 characters, considering the tab as 8 characters (Sorry for still not having a contributor guide). Could you modify this part of code to fit in 80 characters?

agms-cesar replied:
Strange. In my editor is 78 col length this line. it seems to me tab character is 8 spaces here in github

alla3 commented on src/knot_cloud.c:
```diff
@@ -469,11 +421,13 @@ int knot_cloud_unregister_device(const char *id)
-					   json_str);
+	mq_message_data_t mq_message = {
+				 MQ_MESSAGE_TYPE_DIRECT, MQ_EXCHANGE_DEVICE,
+				 MQ_CMD_DEVICE_UNREGISTER, MQ_MSG_EXPIRATION_TIME_MS, json_str,
```
Could you modify this part of code to fit in 80 characters?

agms-cesar replied:
I think the problem is related to github, in my text editor this column length is 78

alla3 commented on src/knot_cloud.c:
```diff
@@ -518,14 +470,12 @@ int knot_cloud_auth_device(const char *id, const char *token)
-					       json_str);
+	mq_message_data_t mq_message = {
+				 MQ_MESSAGE_TYPE_DIRECT_RPC, MQ_EXCHANGE_DEVICE,
+				 MQ_CMD_DEVICE_AUTH, MQ_MSG_EXPIRATION_TIME_MS, json_str,
+				 knot_cloud_events[AUTH_MSG], MQ_DEFAULT_CORRELATION_ID
```
Could you modify this part of code to fit in 80 characters?

alla3 commented on src/knot_cloud.c:
```diff
@@ -568,11 +516,14 @@ int knot_cloud_update_schema(const char *id, struct l_queue *schema_list)
-					   json_str);
+	mq_message_data_t mq_message = {
+				 MQ_MESSAGE_TYPE_DIRECT,
+				 MQ_EXCHANGE_DEVICE, MQ_CMD_SCHEMA_SENT, 
```
Could you remove the boundary white space at the end of the line?
```suggestion
				 MQ_EXCHANGE_DEVICE, MQ_CMD_SCHEMA_SENT,
```

agms-cesar replied:
Yes, i'll do that

alla3 commented on src/mq.c:
```diff
@@ -434,120 +439,76 @@ static int mq_publish_message(const char *exchange,
+	int8_t res;
+	switch (message->msg_type) {
+		case MQ_MESSAGE_TYPE_DIRECT:
+			res = mq_publish(message->exchange, AMQP_EXCHANGE_TYPE_DIRECT,
+					 message->routing_key, headers, num_of_headers,
+					 message->expiration_ms, amqp_empty_bytes, NULL,
```
Could you modify this part of code to fit in 80 characters?

alla3 commented on src/mq.c:
```diff
@@ -434,120 +439,76 @@ static int mq_publish_message(const char *exchange,
+			break;
+		case MQ_MESSAGE_TYPE_DIRECT_RPC:
+			{
+				amqp_bytes_t reply_to = amqp_cstring_bytes(message->reply_to);
+				res = mq_publish(message->exchange, AMQP_EXCHANGE_TYPE_DIRECT,
+					 message->routing_key, headers, num_of_headers,
+					 message->expiration_ms, reply_to,
+					 message->correlation_id, message->body);
```
Could you modify this part of code to fit in 80 characters?

alla3 commented on src/mq.c:
```diff
@@ -434,120 +439,76 @@ static int mq_publish_message(const char *exchange,
+			break;
+			}
+		case MQ_MESSAGE_TYPE_FANOUT:
+			res = mq_publish(message->exchange, AMQP_EXCHANGE_TYPE_FANOUT,
+					 NULL, headers, num_of_headers, message->expiration_ms,
```
Could you modify this part of code to fit in 80 characters?

alla3 commented on src/mq.c:
```diff
@@ -434,120 +439,76 @@ static int mq_publish_message(const char *exchange,
+		case MQ_MESSAGE_TYPE_FANOUT:
+			res = mq_publish(message->exchange, AMQP_EXCHANGE_TYPE_FANOUT,
+					 NULL, headers, num_of_headers, message->expiration_ms,
+				  	 amqp_empty_bytes, NULL, message->body);
```
Could you to remove these extra white spaces between tabs?
```suggestion
					 amqp_empty_bytes, NULL, message->body);
```

alla3 commented on src/mq.c:
```diff
@@ -561,53 +522,49 @@ amqp_bytes_t mq_declare_new_queue(const char *name)
-		return -1;
+	if (current_queue.bytes) {
+			if (mq_ctx.conn) {
+				amqp_queue_delete(mq_ctx.conn, 1, current_queue, 0, 0);
+				if (amqp_get_rpc_reply(mq_ctx.conn).reply_type !=
```
Could you modify this part of code to fit in 80 characters?

alla3 commented on src/mq.c:
```diff
@@ -561,53 +522,49 @@ amqp_bytes_t mq_declare_new_queue(const char *name)
 		l_error("Out of memory while copying queue buffer");
-
-	return queue;
+		
```
Could you remove these boundary tabs?
```suggestion

```

alla3 reviewed (commented):
LGTM. I just made some observations about the code formatting.

CaioCordeiro reviewed (approved):
LGTM.

agms-cesar committed:
Removes amqp_bytes_t dependency from knot_cloud

- Removes queue_fog from knot_cloud.c;
- Creates a worker_queue in mq.c;
- Moves previous queue_fog delete method to mq.c;
- Changes mq_prepare_direct_queue return type to integer;
- Changes mq_declare_new_queue return type to integer;
- Modifies mq_delete_queue signature by removing amqp_bytes_t;
- Modifies mq_consumer_queue signature by removing amqp_bytes_t;
- Adds logic to queue methods to access internal worker_queue.
